optaplanner-examples/src/main/java/org/optaplanner/examples/nurserostering/optional/score/NurseRosteringConstraintProvider.java

```diff
@@ -0,0 +1,480 @@
+                                .filter((employeeContractPair,
+                                        shiftList) -> new EmployeeConsecutiveAssignmentStart(
+                                                employeeContractPair.getLeft(), shiftList.getItems().first())
+                                                        .isWeekendAndNotFirstDayOfWeekend())
+                                .penalize("org.optaplanner.examples.nurserostering.solver", "startOnNotFirstDayOfWeekend",
+                                        HardSoftScore.ONE_SOFT,
+                                        (employeeContractPair, shiftList) -> new EmployeeConsecutiveAssignmentStart(
+                                                employeeContractPair.getLeft(), shiftList.getItems().first())
+                                                        .getDistanceToFirstDayOfWeekend()
+                                                * employeeContractPair.getRight().getWeight());
+    }
+
+    Constraint endOnNotLastDayOfWeekend(ConstraintFactory constraintFactory) {
+        return getConsecutiveShifts(constraintFactory.from(BooleanContractLine.class)
+                .filter(minMaxContractLine -> minMaxContractLine
+                        .getContractLineType() == ContractLineType.COMPLETE_WEEKENDS &&
+                        minMaxContractLine.isEnabled()))
+                                .filter((employeeContractPair,
+                                        shiftList) -> new EmployeeConsecutiveAssignmentEnd(
+                                                employeeContractPair.getLeft(), shiftList.getItems().last())
+                                                        .isWeekendAndNotLastDayOfWeekend())
+                                .penalize("org.optaplanner.examples.nurserostering.solver", "endOnNotLastDayOfWeekend",
+                                        HardSoftScore.ONE_SOFT,
+                                        (employeeContractPair, shiftList) -> new EmployeeConsecutiveAssignmentEnd(
+                                                employeeContractPair.getLeft(), shiftList.getItems().last())
+                                                        .getDistanceToLastDayOfWeekend()
+                                                * employeeContractPair.getRight().getWeight());
+    }
+
+    // Identical shiftTypes during a weekend
+    Constraint identicalShiftTypesDuringWeekend(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(BooleanContractLine.class)
+                .filter(minMaxContractLine -> minMaxContractLine
+                        .getContractLineType() == ContractLineType.IDENTICAL_SHIFT_TYPES_DURING_WEEKEND &&
+                        minMaxContractLine.isEnabled())
+                .join(Employee.class, Joiners.equal(ContractLine::getContract, Employee::getContract))
+                .join(ShiftDate.class, Joiners.filtering((c, e, date) -> date.getDayOfWeek() == DayOfWeek.SUNDAY))
```

Comment:
Considering that the filter is only applied to ShiftDate and nothing else, this will perform better as a nested filtered stream. (See my latest blog.)